controllers/cassandrabackup_controller.go

```diff
@@ -73,6 +75,57 @@ func (r *CassandraBackupReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 
 	backup := instance.DeepCopy()
 
+	// Verify the CassandraBackup has a finalizer
+	if !ctrlutil.ContainsFinalizer(backup, finalizerName) {
+		ctrlutil.AddFinalizer(backup, finalizerName)
+		err = r.Update(ctx, backup)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+		}
+	}
+
+	// GetDeletionTimestamp is set by the apiserver, not us. That's why we need another (DeletionInProgress) progress status
+	if backup.GetDeletionTimestamp() != nil {
+		if backup.Status.DeletionInProgress {
```

Comment:
I am a bit uncomfortable with this check. I fear that we could wind up in a situation like we have with #57. Could we instead query Medusa here to verify whether the backup exists?

Reply:
No, that would do a different thing - this backup.Status is api server matching. Checking if Medusa has it and then proceeding would cause concurrency issue where two (or more) different instances would race to modify the same object suddenly.

Reply:
We would need the generation filtering in the medusa's reconciliation processing. We could also check that the timestamp is not too far in the past (and redo if the deletion has taken more than an hour for example), but I'm not sure what happens if one redeletes it (will the timestamp change).

Reply:
I am referring to specifically to the check at line 89 - `if backup.Status.DeletionInProgress {`. I think this is problematic. 

We patch the status setting `DeletionInProgress` to `true` and then proceed to do the actual work. Suppose the call to `removeBackup` at line 106 fails and then the patch at line 111 to flip `DeletionInProgress` also fails. If this happens I think we get stuck. I don't think we will retry the deletion in subsequent reconciliations.

If we remove the check on `backup.Status.DeletionInProgress` then subsequent reconciliations would retry the delete options. I am not sure if that would cause any problems on the Medusa side.

How about this. Suppose we store a cache of deletion jobs. This could be a simple mapping of `types.NamespacedName`s to deletion status which could just be in progress or failed. I don't think we need to store succeeded. Upon successful completion we just remove the entry from the map. When the deletion goroutine finishes, it updates the cache accordingly. If the deletion is reported as failed, then we would know to retry. 

We currently only use a single goroutine and a single replica of the operator to deployment to perform reconciliations. If we add more goroutines or use multiple replicas, then this doesn't work. I don't see either of those changes happening any time soon though. If at any point we decide to introduce concurrency or parallel processing for medusa-operator reconciliation, then I think a k8s job would be the best solution.

Reply:
It seems like we're really fighting the fact that we can't treat Medusa like the full service that it probably should be.  We're trying to use it as a proxy mechanism for the management of these remote resources when really it would be good if we could delegate that responsibility to it all together, let it maintain the state of what does and doesn't need to be cleaned up and we simply ask that state of it when we want to update the state in the cluster.  

From the convo I had with @jsanda a bit last night, seems like that's just not what Medusa is right now, so while we work towards that better world a local cache option seems reasonable.  Obviously has some downsides with availability of that state, but given the situation and caveats we've got about only having a single replica, I think we can live with it.